mmdet/datasets/coco.py

```diff
@@ -422,30 +425,34 @@ def evaluate(self,
         Returns:
             dict[str, float]: COCO style evaluation metric.
         """
-
-        metrics = metric if isinstance(metric, list) else [metric]
-        allowed_metrics = ['bbox', 'segm', 'proposal', 'proposal_fast']
-        for metric in metrics:
-            if metric not in allowed_metrics:
-                raise KeyError(f'metric {metric} is not supported')
         if iou_thrs is None:
             iou_thrs = np.linspace(
                 .5, 0.95, int(np.round((0.95 - .5) / .05)) + 1, endpoint=True)
         if metric_items is not None:
             if not isinstance(metric_items, list):
                 metric_items = [metric_items]
 
-        result_files, tmp_dir = self.format_results(results, jsonfile_prefix)
-
         eval_results = OrderedDict()
-        cocoGt = self.coco
+        if hasattr(self, 'THING_CLASSES') and hasattr(self, 'STUFF_CLASSES'):
+            cocoGt = COCO(self.ins_ann_file)
+            self.cat_ids = cocoGt.get_cat_ids(cat_names=self.THING_CLASSES)
+            assert self.ins_ann_file is not None, 'Annotation '\
+                'file for instance segmentation or object detection ' \
+                'shuold not be None'
+        else:
+            cocoGt = self.coco
+            self.cat_ids = cocoGt.get_cat_ids(cat_names=self.CLASSES)
+
         for metric in metrics:
             msg = f'Evaluating {metric}...'
             if logger is None:
                 msg = '\n' + msg
             print_log(msg, logger=logger)
 
             if metric == 'proposal_fast':
+                if hasattr(self, 'THING_CLASSES') and hasattr(
+                        self, 'STUFF_CLASSES'):
+                    results = [result['ins_results'] for result in results]
```

Comment:
These two attributes are in the subclass. This logic should not be in the parent class.

Reply:
Element in `results` returned by object detection model or instance segmentation model is `np.array` or `tuple[np.array, np.array]`. `results` returned by panoptic segmentation model is a dict, `results['ins_results']` is `tuple[np.array, np.array]`. so replace ` if hasattr(self, 'THING_CLASSES') and hasattr(self, 'STUFF_CLASSES'):` with ` if isinstance(results[0], dict) and 'ins_results' in results[0]:` to  may be better.

Reply:
And I found that original coco evaluation does not support instance model to do `proposal_fast`. So here, we needn't to consider `proposal_fast` for instance segmentation result returned by a panoptic segmentation model.